File: cpp/src/block.cpp
#include "types.hpp"

#include "util.hpp"

#include <algorithm>

namespace filefs::detail {

std::uint32_t FileSystemState::genblockindex() {
  Block block{};

  if (tmp.new_unused_blockhead > 0U) {
    const std::uint32_t blockindex = tmp.new_unused_blockhead;
    if (!readblock(blockindex, block)) {
      return 0U;
    }
    tmp.new_unused_blockhead = b4_to_u32(std::span<const std::uint8_t, 4>(block.data() + 4, 4));
    return blockindex;
  }

  const std::uint32_t blockindex = tmp.new_total_blocksize;
  const std::uint32_t addindex = blockindex - tmp.total_blocksize;
  const std::uint64_t pos = static_cast<std::uint64_t>(addindex) * static_cast<std::uint64_t>(4 + BlockSize);
  if (!set_pos(tmp.fp_add.get(), pos)) {
    return 0U;
  }

  std::array<std::uint8_t, 4> b4{};
  u32_to_b4(blockindex, b4);
  if (!write_exact(tmp.fp_add.get(), b4.data(), b4.size()) ||
      !write_exact(tmp.fp_add.get(), block.data(), block.size())) {
    return 0U;
  }
  ++tmp.new_total_blocksize;
  return blockindex;
}

bool FileSystemState::readblock(const std::uint32_t blockindex, Block& block) {
  std::uint64_t pos = static_cast<std::uint64_t>(blockindex) * BlockSize;
  if (!set_pos(fp.get(), pos)) {
    return false;
  }

  std::array<std::uint8_t, 4> buf{};
  if (!read_exact(fp.get(), buf.data(), buf.size())) {
    if (tmp.state == 0U || blockindex < tmp.total_blocksize) {
      return false;
    }
    const std::uint32_t addindex = blockindex - tmp.total_blocksize;
    pos = static_cast<std::uint64_t>(addindex) * static_cast<std::uint64_t>(BlockSize + 4) + 4U;
    if (!set_pos(tmp.fp_add.get(), pos)) {
      return false;
    }
    return read_exact(tmp.fp_add.get(), block.data(), block.size());
  }

  if (tmp.state == 0U) {
    std::copy(buf.begin(), buf.end(), block.begin());
    return read_exact(fp.get(), block.data() + 4, block.size() - 4);
  }

  std::uint32_t cpindex = b4_to_u32(buf);
  pos = static_cast<std::uint64_t>(cpindex) * static_cast<std::uint64_t>(BlockSize + 4);
  if (!set_pos(tmp.fp_cp.get(), pos)) {
    return false;
  }

  std::array<std::uint8_t, 4> b4{};
  if (!read_exact(tmp.fp_cp.get(), b4.data(), b4.size())) {
    std::copy(buf.begin(), buf.end(), block.begin());
    return read_exact(fp.get(), block.data() + 4, block.size() - 4);
  }
  if (b4_to_u32(b4) != blockindex) {
    std::copy(buf.begin(), buf.end(), block.begin());
    return read_exact(fp.get(), block.data() + 4, block.size() - 4);
  }
  return read_exact(tmp.fp_cp.get(), block.data(), block.size());
}

bool FileSystemState::writeblock(const std::uint32_t blockindex, const Block& block) {
  if (tmp.state == 0U) {
    return false;
  }

  std::uint64_t pos = static_cast<std::uint64_t>(blockindex) * BlockSize;
  if (!set_pos(fp.get(), pos)) {
    return false;
  }

  std::array<std::uint8_t, 4> buf{};
  if (!read_exact(fp.get(), buf.data(), buf.size())) {
    if (blockindex < tmp.total_blocksize) {
      return false;
    }
    const std::uint32_t addindex = blockindex - tmp.total_blocksize;
    pos = static_cast<std::uint64_t>(addindex) * static_cast<std::uint64_t>(BlockSize + 4) + 4U;
    if (!set_pos(tmp.fp_add.get(), pos)) {
      return false;
    }
    return write_exact(tmp.fp_add.get(), block.data(), block.size());
  }

  std::array<std::uint8_t, 4> b4{};
  std::uint32_t cpindex = b4_to_u32(buf);
  pos = static_cast<std::uint64_t>(cpindex) * static_cast<std::uint64_t>(BlockSize + 4);
  if (!set_pos(tmp.fp_cp.get(), pos)) {
    return false;
  }

  if (!read_exact(tmp.fp_cp.get(), b4.data(), b4.size())) {
    cpindex = tmp.cp_size;
    pos = static_cast<std::uint64_t>(cpindex) * static_cast<std::uint64_t>(BlockSize + 4);
    if (!set_pos(tmp.fp_cp.get(), pos)) {
      return false;
    }
    u32_to_b4(blockindex, b4);
    if (!write_exact(tmp.fp_cp.get(), b4.data(), b4.size()) ||
        !write_exact(tmp.fp_cp.get(), block.data(), block.size())) {
      return false;
    }

    pos = static_cast<std::uint64_t>(blockindex) * BlockSize;
    if (!set_pos(fp.get(), pos)) {
      return false;
    }
    u32_to_b4(cpindex, b4);
    if (!write_exact(fp.get(), b4.data(), b4.size())) {
      return false;
    }
    ++tmp.cp_size;
    return true;
  }

  if (b4_to_u32(b4) != blockindex) {
    cpindex = tmp.cp_size;
    pos = static_cast<std::uint64_t>(cpindex) * static_cast<std::uint64_t>(BlockSize + 4);
    if (!set_pos(tmp.fp_cp.get(), pos)) {
      return false;
    }
    u32_to_b4(blockindex, b4);
    if (!write_exact(tmp.fp_cp.get(), b4.data(), b4.size()) ||
        !write_exact(tmp.fp_cp.get(), block.data(), block.size())) {
      return false;
    }

    pos = static_cast<std::uint64_t>(blockindex) * BlockSize;
    if (!set_pos(fp.get(), pos)) {
      return false;
    }
    u32_to_b4(cpindex, b4);
    if (!write_exact(fp.get(), b4.data(), b4.size())) {
      return false;
    }
    ++tmp.cp_size;
    return true;
  }

  pos = static_cast<std::uint64_t>(cpindex) * static_cast<std::uint64_t>(BlockSize + 4) + 4U;
  if (!set_pos(tmp.fp_cp.get(), pos)) {
    return false;
  }
  return write_exact(tmp.fp_cp.get(), block.data(), block.size());
}

bool FileSystemState::removeblock(const std::uint32_t blockindex) {
  if (tmp.state == 0U) {
    return false;
  }

  std::uint64_t pos = static_cast<std::uint64_t>(blockindex) * BlockSize;
  if (!set_pos(fp.get(), pos)) {
    return false;
  }

  std::array<std::uint8_t, 4> buf{};
  if (!read_exact(fp.get(), buf.data(), buf.size())) {
    if (blockindex < tmp.total_blocksize) {
      return false;
    }
    const std::uint32_t addindex = blockindex - tmp.total_blocksize;
    pos = static_cast<std::uint64_t>(addindex) * static_cast<std::uint64_t>(BlockSize + 4) + 8U;
    if (!set_pos(tmp.fp_add.get(), pos)) {
      return false;
    }
    std::array<std::uint8_t, 4> b4{};
    u32_to_b4(tmp.new_unused_blockhead, b4);
    if (!write_exact(tmp.fp_add.get(), b4.data(), b4.size())) {
      return false;
    }
    tmp.new_unused_blockhead = blockindex;
    return true;
  }

  std::array<std::uint8_t, 4> b4{};
  std::uint32_t cpindex = b4_to_u32(buf);
  pos = static_cast<std::uint64_t>(cpindex) * static_cast<std::uint64_t>(BlockSize + 4);
  if (!set_pos(tmp.fp_cp.get(), pos)) {
    return false;
  }

  if (!read_exact(tmp.fp_cp.get(), b4.data(), b4.size())) {
    cpindex = tmp.cp_size;
    pos = static_cast<std::uint64_t>(cpindex) * static_cast<std::uint64_t>(BlockSize + 4);
    if (!set_pos(tmp.fp_cp.get(), pos)) {
      return false;
    }
    u32_to_b4(blockindex, b4);
    if (!write_exact(tmp.fp_cp.get(), b4.data(), b4.size())) {
      return false;
    }
    Block block{};
    u32_to_b4(tmp.new_unused_blockhead, std::span<std::uint8_t, 4>(block.data(), 4));
    if (!write_exact(tmp.fp_cp.get(), block.data(), block.size())) {
      return false;
    }

    pos = static_cast<std::uint64_t>(blockindex) * BlockSize;
    if (!set_pos(fp.get(), pos)) {
      return false;
    }
    u32_to_b4(cpindex, b4);
    if (!write_exact(fp.get(), b4.data(), b4.size())) {
      return false;
    }
    ++tmp.cp_size;
    tmp.new_unused_blockhead = blockindex;
    return true;
  }

  if (b4_to_u32(b4) != blockindex) {
    cpindex = tmp.cp_size;
    pos = static_cast<std::uint64_t>(cpindex) * static_cast<std::uint64_t>(BlockSize + 4);
    if (!set_pos(tmp.fp_cp.get(), pos)) {
      return false;
    }
    u32_to_b4(blockindex, b4);
    if (!write_exact(tmp.fp_cp.get(), b4.data(), b4.size())) {
      return false;
    }
    Block block{};
    u32_to_b4(tmp.new_unused_blockhead, std::span<std::uint8_t, 4>(block.data(), 4));
    if (!write_exact(tmp.fp_cp.get(), block.data(), block.size())) {
      return false;
    }

    pos = static_cast<std::uint64_t>(blockindex) * BlockSize;
    if (!set_pos(fp.get(), pos)) {
      return false;
    }
    u32_to_b4(cpindex, b4);
    if (!write_exact(fp.get(), b4.data(), b4.size())) {
      return false;
    }
    ++tmp.cp_size;
    tmp.new_unused_blockhead = blockindex;
    return true;
  }

  pos += 8U;
  if (!set_pos(tmp.fp_cp.get(), pos)) {
    return false;
  }
  u32_to_b4(tmp.new_unused_blockhead, b4);
  if (!write_exact(tmp.fp_cp.get(), b4.data(), b4.size())) {
    return false;
  }
  tmp.new_unused_blockhead = blockindex;
  return true;
}

std::uint32_t FileSystemState::find_path_blockindex(std::uint32_t blockindex, const std::string_view pathname) {
  Block block{};
  std::uint32_t index = blockindex;
  std::array<std::uint8_t, 15> s{};

  if (!readblock(index, block)) {
    return 0U;
  }
  const std::uint32_t stop_blockindex =
      b4_to_u32(std::span<const std::uint8_t, 4>(block.data() + 12 + 1 + 14 + 4, 4));
  const std::uint16_t offset =
      b2_to_u16(std::span<const std::uint8_t, 2>(block.data() + 12 + 1 + 14 + 4 + 4, 2));

  for (;;) {
    int k = BlockHead;
    for (int i = 0; i < BlockItemMaxCount; ++i) {
      const std::uint8_t state = block[static_cast<std::size_t>(k)];
      ++k;
      const std::uint8_t dir_file = state & 0x01U;
      if (dir_file == 1U) {
        k += 24;
        if (index == stop_blockindex && k + 1 >= static_cast<int>(offset)) {
          return 0U;
        }
        continue;
      }

      std::copy_n(block.data() + k, BlockNameMaxSize, s.data());
      k += static_cast<int>(BlockNameMaxSize);
      if (fixed_cstr(s) != pathname) {
        k += 10;
        if (index == stop_blockindex && k + 1 >= static_cast<int>(offset)) {
          return 0U;
        }
        continue;
      }
      return b4_to_u32(std::span<const std::uint8_t, 4>(block.data() + k, 4));
    }

    index = b4_to_u32(std::span<const std::uint8_t, 4>(block.data() + 4, 4));
    if (index == 0U) {
      return 0U;
    }
    if (!readblock(index, block)) {
      return 0U;
    }
  }
}

}  // namespace filefs::detail

File: zig/src/block.zig
const std = @import("std");
const types = @import("types.zig");
const util = @import("util.zig");

pub fn genblockindex(self: anytype) u32 {
    var block = util.blockZero();

    if (self.tmp.new_unused_blockhead > 0) {
        const blockindex = self.tmp.new_unused_blockhead;
        if (!readblock(self, blockindex, block[0..])) {
            return 0;
        }
        self.tmp.new_unused_blockhead = util.b4ToU32(block[4..8]);
        return blockindex;
    }

    const blockindex = self.tmp.new_total_blocksize;
    const addindex = blockindex - self.tmp.total_blocksize;
    const pos = @as(u64, addindex) * @as(u64, 4 + types.BlockSize);
    const fp_add = self.tmp.fp_add orelse return 0;
    var b4: [4]u8 = undefined;
    util.u32ToB4(blockindex, &b4);
    util.writeAllAt(self.io, fp_add, pos, b4[0..]) catch return 0;
    util.writeAllAt(self.io, fp_add, pos + 4, block[0..]) catch return 0;
    self.tmp.new_total_blocksize += 1;
    return blockindex;
}

pub fn readblock(self: anytype, blockindex: u32, block: []u8) bool {
    const main = self.fp orelse return false;
    const pos = @as(u64, blockindex) * @as(u64, types.BlockSize);
    var buf: [4]u8 = undefined;
    util.readExactAt(self.io, main, pos, buf[0..]) catch {
        if (self.tmp.state == 0) {
            return false;
        }
        if (blockindex < self.tmp.total_blocksize) {
            return false;
        }
        const fp_add = self.tmp.fp_add orelse return false;
        const addindex = blockindex - self.tmp.total_blocksize;
        const add_pos = @as(u64, addindex) * @as(u64, types.BlockSize + 4) + 4;
        util.readExactAt(self.io, fp_add, add_pos, block[0..types.BlockSize]) catch return false;
        return true;
    };

    if (self.tmp.state == 0) {
        @memcpy(block[0..4], buf[0..]);
        util.readExactAt(self.io, main, pos + 4, block[4..types.BlockSize]) catch return false;
        return true;
    }

    var b4: [4]u8 = undefined;
    @memcpy(b4[0..], buf[0..]);
    const cpindex = util.b4ToU32(b4[0..]);
    const fp_cp = self.tmp.fp_cp orelse return false;
    const cp_pos = @as(u64, cpindex) * @as(u64, types.BlockSize + 4);
    util.readExactAt(self.io, fp_cp, cp_pos, b4[0..]) catch {
        @memcpy(block[0..4], buf[0..]);
        util.readExactAt(self.io, main, pos + 4, block[4..types.BlockSize]) catch return false;
        return true;
    };
    const orgindex = util.b4ToU32(b4[0..]);
    if (orgindex != blockindex) {
        @memcpy(block[0..4], buf[0..]);
        util.readExactAt(self.io, main, pos + 4, block[4..types.BlockSize]) catch return false;
        return true;
    }
    util.readExactAt(self.io, fp_cp, cp_pos + 4, block[0..types.BlockSize]) catch return false;
    return true;
}

pub fn writeblock(self: anytype, blockindex: u32, block: []const u8) bool {
    if (self.tmp.state == 0) {
        return false;
    }

    const main = self.fp orelse return false;
    const pos = @as(u64, blockindex) * @as(u64, types.BlockSize);
    var buf: [4]u8 = undefined;
    util.readExactAt(self.io, main, pos, buf[0..]) catch {
        if (blockindex < self.tmp.total_blocksize) {
            return false;
        }
        const addindex = blockindex - self.tmp.total_blocksize;
        const add_pos = @as(u64, addindex) * @as(u64, types.BlockSize + 4) + 4;
        const fp_add = self.tmp.fp_add orelse return false;
        util.writeAllAt(self.io, fp_add, add_pos, block[0..types.BlockSize]) catch return false;
        return true;
    };

    var b4: [4]u8 = undefined;
    @memcpy(b4[0..], buf[0..]);
    var cpindex = util.b4ToU32(b4[0..]);
    const fp_cp = self.tmp.fp_cp orelse return false;
    var cp_pos = @as(u64, cpindex) * @as(u64, types.BlockSize + 4);
    util.readExactAt(self.io, fp_cp, cp_pos, b4[0..]) catch {
        cpindex = self.tmp.cp_size;
        cp_pos = @as(u64, cpindex) * @as(u64, types.BlockSize + 4);
        util.u32ToB4(blockindex, &b4);
        util.writeAllAt(self.io, fp_cp, cp_pos, b4[0..]) catch return false;
        util.writeAllAt(self.io, fp_cp, cp_pos + 4, block[0..types.BlockSize]) catch return false;
        util.u32ToB4(cpindex, &b4);
        util.writeAllAt(self.io, main, pos, b4[0..]) catch return false;
        self.tmp.cp_size +%= 1;
        return true;
    };
    const orgindex = util.b4ToU32(b4[0..]);
    if (orgindex != blockindex) {
        cpindex = self.tmp.cp_size;
        cp_pos = @as(u64, cpindex) * @as(u64, types.BlockSize + 4);
        util.u32ToB4(blockindex, &b4);
        util.writeAllAt(self.io, fp_cp, cp_pos, b4[0..]) catch return false;
        util.writeAllAt(self.io, fp_cp, cp_pos + 4, block[0..types.BlockSize]) catch return false;
        util.u32ToB4(cpindex, &b4);
        util.writeAllAt(self.io, main, pos, b4[0..]) catch return false;
        self.tmp.cp_size +%= 1;
        return true;
    }

    util.writeAllAt(self.io, fp_cp, cp_pos + 4, block[0..types.BlockSize]) catch return false;
    return true;
}

pub fn removeblock(self: anytype, blockindex: u32) bool {
    if (self.tmp.state == 0) {
        return false;
    }

    const main = self.fp orelse return false;
    const pos = @as(u64, blockindex) * @as(u64, types.BlockSize);
    var buf: [4]u8 = undefined;
    util.readExactAt(self.io, main, pos, buf[0..]) catch {
        if (blockindex < self.tmp.total_blocksize) {
            return false;
        }
        const addindex = blockindex - self.tmp.total_blocksize;
        const add_pos = @as(u64, addindex) * @as(u64, types.BlockSize + 4) + 8;
        const fp_add = self.tmp.fp_add orelse return false;
        var b4: [4]u8 = undefined;
        util.u32ToB4(self.tmp.new_unused_blockhead, &b4);
        util.writeAllAt(self.io, fp_add, add_pos, b4[0..]) catch return false;
        self.tmp.new_unused_blockhead = blockindex;
        return true;
    };

    var b4: [4]u8 = undefined;
    @memcpy(b4[0..], buf[0..]);
    var cpindex = util.b4ToU32(b4[0..]);
    const fp_cp = self.tmp.fp_cp orelse return false;
    var cp_pos = @as(u64, cpindex) * @as(u64, types.BlockSize + 4);
    util.readExactAt(self.io, fp_cp, cp_pos, b4[0..]) catch {
        cpindex = self.tmp.cp_size;
        cp_pos = @as(u64, cpindex) * @as(u64, types.BlockSize + 4);
        util.u32ToB4(blockindex, &b4);
        util.writeAllAt(self.io, fp_cp, cp_pos, b4[0..]) catch return false;
        var block = util.blockZero();
        util.u32ToB4(self.tmp.new_unused_blockhead, &b4);
        @memcpy(block[0..4], b4[0..]);
        util.writeAllAt(self.io, fp_cp, cp_pos + 4, block[0..]) catch return false;
        util.u32ToB4(cpindex, &b4);
        util.writeAllAt(self.io, main, pos, b4[0..]) catch return false;
        self.tmp.cp_size +%= 1;
        self.tmp.new_unused_blockhead = blockindex;
        return true;
    };
    const orgindex = util.b4ToU32(b4[0..]);
    if (orgindex != blockindex) {
        cpindex = self.tmp.cp_size;
        cp_pos = @as(u64, cpindex) * @as(u64, types.BlockSize + 4);
        util.u32ToB4(blockindex, &b4);
        util.writeAllAt(self.io, fp_cp, cp_pos, b4[0..]) catch return false;
        var block = util.blockZero();
        util.u32ToB4(self.tmp.new_unused_blockhead, &b4);
        @memcpy(block[0..4], b4[0..]);
        util.writeAllAt(self.io, fp_cp, cp_pos + 4, block[0..]) catch return false;
        util.u32ToB4(cpindex, &b4);
        util.writeAllAt(self.io, main, pos, b4[0..]) catch return false;
        self.tmp.cp_size +%= 1;
        self.tmp.new_unused_blockhead = blockindex;
        return true;
    }

    util.u32ToB4(self.tmp.new_unused_blockhead, &b4);
    util.writeAllAt(self.io, fp_cp, cp_pos + 8, b4[0..]) catch return false;
    self.tmp.new_unused_blockhead = blockindex;
    return true;
}

pub fn findPathBlockindex(self: anytype, blockindex: u32, pathname: []const u8) u32 {
    var block = util.blockZero();
    var index = blockindex;
    var b4: [4]u8 = undefined;
    var b2: [2]u8 = undefined;
    var s: [types.BlockNameMaxSize + 1]u8 = [_]u8{0} ** (types.BlockNameMaxSize + 1);

    if (!readblock(self, index, block[0..])) {
        return 0;
    }
    @memcpy(b4[0..], block[12 + 1 + 14 + 4 .. 12 + 1 + 14 + 8]);
    const stop_blockindex = util.b4ToU32(b4[0..]);
    @memcpy(b2[0..], block[12 + 1 + 14 + 4 + 4 .. 12 + 1 + 14 + 4 + 6]);
    const offset = util.b2ToU16(b2[0..]);

    while (true) {
        var k: usize = types.BlockHead;
        var i: usize = 0;
        while (i < types.BlockItemMaxCount) : (i += 1) {
            const state = block[k];
            k += 1;
            const dir_file = state & 0x01;
            if (dir_file == 1) {
                k += 24;
                if (index == stop_blockindex and k + 1 >= util.u16ToUsize(offset)) {
                    return 0;
                }
                continue;
            }
            @memset(s[0..], 0);
            @memcpy(s[0..types.BlockNameMaxSize], block[k .. k + types.BlockNameMaxSize]);
            k += types.BlockNameMaxSize;
            if (!std.mem.eql(u8, util.fixedCStr(s[0..]), pathname)) {
                k += 10;
                if (index == stop_blockindex and k + 1 >= util.u16ToUsize(offset)) {
                    return 0;
                }
                continue;
            }
            @memcpy(b4[0..], block[k .. k + 4]);
            return util.b4ToU32(b4[0..]);
        }

        @memcpy(b4[0..], block[4..8]);
        index = util.b4ToU32(b4[0..]);
        if (index == 0) {
            return 0;
        }
        if (!readblock(self, index, block[0..])) {
            return 0;
        }
    }
}
